### PR TITLE
DynamoDB Support

### PR DESCRIPTION
**Pull request by johnewart**
Opening because @catsby asked that I open a PR to let folks know I'm working on this and in case anyone has some feedback. Under :construction: - I'll be writing tests and docs as well as finishing off a few things but it is mostly functional at this point.

**Remaining:**

- [x] tests
- [x] docs
- [x] a few things

### CODE REVIEW

**Comment by catsby** [comment]
Thanks @johnewart! I updated your comment with a fancy checklist to keep everyone up to speed on our progress here. Anything I can work on?

**Comment by johnewart** [comment]
Hmm, looks like I'll need to clean up this patch after I finish docs today.

File: builtin/providers/aws/resource_aws_dynamodb_table.go [hunk 1]
```diff
@@ -0,0 +1,610 @@
...
+		newAttributeSet := n.(*schema.Set)
+		for _, attribute := range newAttributeSet.List() {
+			attr := attribute.(map[string]interface{})
+			newAttributes = append(newAttributes, &dynamodb.AttributeDefinition{
+				AttributeName: aws.String(attr["name"].(string)),
+				AttributeType: aws.String(attr["type"].(string)),
+			})
+		}
+
+		req.AttributeDefinitions = newAttributes
```
Comment (catsby) [inline builtin/providers/aws/resource_aws_dynamodb_table.go]: There's no call to `dynamodbconn.UpdateTable(req)` with this locally scoped `req`, or am I missing that? The update test doesn't check any changes in the attribute map so I think this snuck through. 

Alternatively, is it worth building up a single `UpdateTableInput` struct for a single `UpdateTable` call?

**Comment by catsby** [comment]
One question, other than that :+1:

**Comment by johnewart** [comment]
This is actually intentional; the request object gets reused all the way down. Because each update (creation / update of a GSI) will need to ensure that the newly described attributes are provided in the request, it makes more sense to bundle them all in there and then pass them in every subsequent update. Each GSI creation / deletion / update re-uses the request context but just swaps out the list of updates to apply. The alternative would be to figure out all the attributes required for each GSI update / create and then re-build the entire request but it would yield the same effect and be a lot more work.

**Comment by catsby** [comment]
Excellent work @johnewart, thanks!

**Pull request merged**